arrayfire/array_api/array_object.py

```diff
@@ -0,0 +1,1093 @@
+from __future__ import annotations
+
+import array as py_array
+import ctypes
+import enum
+from dataclasses import dataclass
+from typing import Any, List, Optional, Tuple, Union
+
+# TODO replace imports from original lib with refactored ones
+from arrayfire import backend, safe_call
+from arrayfire.algorithm import count
+from arrayfire.array import _get_indices, _in_display_dims_limit
+
+from .device import PointerSource
+from .dtypes import CShape, Dtype
+from .dtypes import bool as af_bool
+from .dtypes import c_dim_t
+from .dtypes import complex64 as af_complex64
+from .dtypes import complex128 as af_complex128
+from .dtypes import float32 as af_float32
+from .dtypes import float64 as af_float64
+from .dtypes import int64 as af_int64
+from .dtypes import supported_dtypes
+from .dtypes import uint64 as af_uint64
+
+ShapeType = Tuple[int, ...]
+# HACK, TODO replace for actual bcast_var after refactoring ~ https://github.com/arrayfire/arrayfire/pull/2871
+_bcast_var = False
+
+# TODO use int | float in operators -> remove bool | complex support
+
+
+@dataclass
+class _ArrayBuffer:
+    address: Optional[int] = None
+    length: int = 0
+
+
+class Array:
+    def __init__(
```

Comment:
This PR ought to describe a new structure with fewer changes to ArrayFire API. First of many, tbh :)
Changing init now will affect a major part of this PR and follow-up #262. Can we postpone these changes (that are crucial I agree) to the third one, so as to avoid huge code merge issues? Proposed changes in #262 will simplify the implementation of `asarray` method, thus I wanted to keep creation functions (https://data-apis.org/array-api/latest/API_specification/creation_functions.html) till the finished backend.